### PR TITLE
Ssr merge

### PR DESCRIPTION
**Pull request by maximedenes**
This PR makes math-comp use the ssreflect plugin that is being merged in Coq. I had to disable a test on primitive projection unfolding, will need to investigate if there's a bug.

### CODE REVIEW

File: mathcomp/ssreflect/Makefile.coq-makefile [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
-		EXTRA="ssrmatching.mli ssrmatching.ml4 ssrmatching.v";\
+		EXTRA="ssrmatching.mli ssrmatching.ml4 ssrmatching.v ssreflect.ml4";\
 		MLLIB=ssreflect_plugin.mllib;\
+		$$LN $(1)/plugin/$(V)/ssreflect.ml4 .;\
 	;;\
-	*)\
+	v8.6*)\
 		$$LN $(1)/plugin/$(V)/ssreflect_plugin.mlpack .;\
+		$$LN $(1)/plugin/$(V)/ssreflect.ml4 .;\
+		MLLIB=ssreflect_plugin.mlpack;\
```
Comment (gares) [inline mathcomp/ssreflect/Makefile.coq-makefile]: no EXTRA here? that is probably a bug

Comment (maximedenes) [inline mathcomp/ssreflect/Makefile.coq-makefile]: It is surprising, but was already like that before my PR, right? Do you confirm it's a bug?

**Comment by maximedenes** [comment]
@gares Do you have a suggestion on how to restore 8.6 compatibility at the `.v` level? The current setup requires a `From Coq Require Import ssreflect`.

**Comment by ejgallego** [comment]
@maximedenes you could play `LoadPath` tricks, but that won't be fun; otherwise I guess you would need to have a different `ssreflect.v` for 8.5/8.6 and a different one for `master`, that just exports the Coq shipped module.

**Comment by ejgallego** [comment]
There is an even better option that is to revert to the `mathcomp` prefix, I cannot stress how annoying is now that all my ssreflect examples have to come in two versions...

**Comment by maximedenes** [comment]
> There is an even better option that is to revert to the mathcomp prefix

Isn't that the same as your loadpath suggestion? I didn't change the prefix in any way, it is just the fact that we moved the files that had this side effect. To compensate for it, we should tweak the loadpath I guess.

**Comment by maximedenes** [comment]
Btw, can a prefix be bound to several physical paths?

**Comment by ejgallego** [comment]
> Isn't that the same as your loadpath suggestion?

The prefix is set arbitrarily by Coq, (in fact in SerAPI/jsCoq I have to imitate that arbitrary behavior).

You could well have a rule that declares `plugins/ssr` to be under prefix `mathcomp`. It is a bit uglier, but compatible.

**Comment by ejgallego** [comment]
> Btw, can a prefix be bound to several physical paths?

it will complain with a warning, `overriding-logical-loadpath` whether it does make sense, I don't know.

**Comment by maximedenes** [comment]
So I guess I don't understand your suggestion. You'd loose access to the `mathcomp` libraries, right?

**Comment by ejgallego** [comment]
Likely you would have to put the full `mathcomp.ssreflect` in Coq, the rest should fine, right?

**Comment by maximedenes** [comment]
I don't follow. If the `mathcomp` prefix points to `plugins/ssr`, `From mathcomp Require Import ssralg` will fail. If it points to the place where mathcomp is installed, it is `Require Import mathcomp.ssreflect` that will fail, right?

**Comment by ejgallego** [comment]
The `mathcomp.ssreflect` prefix would point to `plugins/ssr`, `From mathcomp Require Import ssralg` resolves to `mathcomp.algebra.ssralg` I believe.

`Require Import mathcomp.ssreflect` already fails in 8.6 as far as I can see; it makes no sense with the current organization.

**Comment by ejgallego** [comment]
Anyways the current loadpath story seems kinda bogus, the main reason duplication is not allowed is that `find_load_path` can determine the logical path a module uses in `start_library`.

IMHO this is the wrong approach, as it should be the user of `start_library` who provides that and I even believe that I saw some bug / non-flexible behavior due to that.

I'm afraid that the whole story with `start/declare_library` is a mess now, including real bugs such as 5556; I  am close to have a first patch ready thou to attack 5556 that in fact will properly pack together libraries with their loadpaths and other metaproperties that are set in an inconsistent way today. This will be done by enriching `Stm.init` so in fact the STM becomes in full control of the notion of "document/compilation unit".

Lets see where we arrive.

**Comment by maximedenes** [comment]
> The mathcomp.ssreflect prefix would point to plugins/ssr

Ok, now I understand. There will still be an incompatibility for, e.g. `Require Import mathcomp.ssreflect.finfun` and all other files there. But it was recommended to do `From mathcomp Require Import finfun`, and that should still work.

Where do you think I should add this binding? Just as an ad-hoc thing in `coqinit.ml`, like it is done for the `Coq` prefix?

**Comment by ejgallego** [comment]
> But it was recommended to do From mathcomp Require Import finfun, and that should still work.

How would that work now?

**Comment by maximedenes** [comment]
I thought your suggestion was to bind the `mathcomp.ssreflect` prefix, but not the `mathcomp` one.

**Comment by ejgallego** [comment]
Unfortunately my suggestion does require moving the whole of `ssreflect` into Coq with the current state of affairs, what would you bind `mathcomp` to?

So I thought indeed the original plan here was to provide a different version of `ssreflect.v` for 8.6/8.7 . IMHO that would be fine, even if we have the pain to have users to install mathcomp even for just the tactics. Well, now you'll have to provide also different versions of `ssrbool` etc... Pity we still don't have cpp support in coq_makefile.

The only other alternative I see is to move the full `ssreflect` prefix into `plugins/ssr`; this was my original idea so people could just use the thing without any extra effort in a compatible way.

I think other alternatives will require rework of the loadpath mode, or adding cpp support to coq_makefile.

**Pull request merged**